Abstract CLPR Connector Protocol (#12)
edward-swirldslabs/hedera-improvement-proposal: HIP/hip-0000-cross-ledger-protocol-clpr.md

its state for later use in verification of the message payload. CLPR can
be viewed as _lighter than light clients_ because the state proofs of one
ledger are not added to the state of another ledger. Light clients also do
not have provisions in their protocol for specifying who pays for the cost
of validating the state proofs and handling the messages on the receiving
network. With CLPR, the entity registering the connector determines the
payment mechanics for message handling on each endpoint.

## Personas and User Stories

Personas

* _Connector Financier_: Someone willing to finance interledger interactions
* _dApp Developer_: A developer of interledger applications.
* _End User_: A person who uses intereldger applications.

User Stories

* As a Connector Financier, I want to setup connectors between ledgers and
  make financial arrangements with dApp developers to support the execution
  of their interledger applications.
* As a dApp developer, I want to create an interledger dApp which will
  employ specific connectors between specific ledgers.
* As a user of dApps, from my account on one ledger, I want to invoke a
  specific function or capability on another ledger.

> 

Reviewer: We may want to add some use cases here for operators of private ledgers who wish to connect their ledgers to a public network _while maintaining privacy_ for transactions on the private ledger.

Author: I'll add a todo to fill this out.

types. When one endpoint initiates a network connection to a remote
endpoint, the initiating endpoint must use the network messaging paradigm of
the receiving endpoint. For example, suppose endpoint A uses gRPC as is
network messaging paradigm with protobuf as its data format for state
proofs and endpoint B uses a custom bytes format for network messages and
representing state proofs. If B initiates the connection to A, it must use
gRPC and communicate its state proofs to A in a raw sequence of bytes within
the context of a protobuf message. Endpoint A needs know how to parse the
bytes to recover the content of B's state proofs. Similarly, if A
initiates the connection to B, it must use the custom network messaging
format that B uses and A must provide its protobuf state proofs as a byte
serialized payload in B's network messaging format. B must know how to
parse the protobuf payload to recover the content of A's state proofs.

#### Abstract Connector Registration

At minimum, the following connector metadata will be included in the
connector registration or update transaction and is signed by the connector
private key:

1. `connector_key` : the connector's public key for pairing

> 2. `local_clpr_endpoint_id` : the identifier of the local clpr endpoint
> 3. `local_clpr_endpoint_type` : indicates the type of local clpr endpoint
> 4. `local_clpr_endpoint_config` : local endpoint configuration
> 5. `remote_clpr_endpoint_type` : indicatest the type of remote clpr endpoint
> 6. `remote_clpr_endpoint_config` : remote endpoint configuration for that type

Reviewer: Given that ID, Type, and Config are present for both local and remote, would it make sense to just have a single flag indicating local/remote, and then the three common fields are all the same?
```suggestion
1. `is_endpoint_remote` : Flag set true if the endpoint is remote,
                          or false if the endpoint is local
2. `endpoint_id` : the identifier of the clpr endpoint
3. `endpoint_type` : indicates the type of clpr endpoint
4. `endpoint_config` : endpoint configuration
```

    1. `clpr_endpoint_id` : the unique identifier for the CLPR Endpoint

>     2. `clpr_endpoint_ips` : an optional list of CLPR Endpoint IP addresses

Reviewer: Might want to rename these to "addresses" and allow for a bit more flexibility of address structure.  IP is only one of several mechanisms, including FQDN, for identifying an endpoint.

Author: changing to addresses.

>     2. `clpr_endpoint_ips` : an optional list of CLPR Endpoint IP addresses

Reviewer 2: should there be ports as well as IPs or are ports standardized?

Author: good catch, I'll add port

through updates to the connector or additional connector registrations. If the
list of ip addresses for an endpoint is empty in all connector registrations,
the local ledger will only receive connections from the remote ledger. If a
pair of CLPR Endpoints has a connector registered between them, at least one of
the CLPR Endpoints must be configured with the IP Addresses of the other
endpoint otherwise neither endpoint will attempt to communicate to the other.

#### Abstract CLPR Channel Protocol

When one CLPR Endpoint (local) reaches out to a remote CLPR Endpoint, there
is at least 1 registered connector on the local endpoint that has provided
one or more initial IP addresses to connect to. It is assumed that all IP
addresses in a connector registration reach the same CLPR Endpoint. Once two
CLPR Endpoints have established a channel of communication between them, the
configuration for the remote endpoints is updated through the endpoints
directly talking to each other through a channel. The process of
communicating through a channel between two CLPR Endpoints to exchange connector
state and queued messages is as follows:

> 1. The initiating endpoint sends a `Proof of Remote CLPR Endpoint Config`

Reviewer 2: how does this map to GRPC? seems like it is a series of calls, would be nice if it could be single call with all data and getting all data needed as a response. Or could be streaming. Just feels little wrong to have to orchestrate an exact flow of stateful calls.

Author: This could be broken up into two services. 
1. Proof Configuration Service
2. Connector State Service. 

Before sending state proofs of connector content, each endpoint needs to know the applicable throttle settings for the content and the state proof of the remote connector's state to know what messages it needs.  

I'm non-attached and agree that the Hiero gRPC implementation should be what makes the most intuitive sense.  We should discuss this on our call.

   channel message which is a state proof showing the initiating endpoint's
   configuration for the receiving endpoint.
    1. In response, the receiving endpoint provides state proofs which
       attest to the receiving endpoint's latest configuration.
2. The receiving endpoint sends a `Proof of Remote CLPR Endpoint Config`
   channel message which is a state proof showing the receiving endpoint's
   configuration for the initiating endpoint.
    1. In response, the initiating endpoint provides state proofs which

>        attest to the initiating endpoint's latest configuration.

Reviewer: The sub bullet on each of these seems to refer to the cross-connect.
That is, on 1 the sub bullet appears to refer to (2), and the sub bullet on 2 appears to refer to (1).

Author: I believe the sub-bullet is correct.  In the parent bullet, one endpoint sends the config it has for the remote endpoint.  In the sub-bullet, the remote endpoint responds with the updated config.  
* Request: `Here's the config I have for you`
   * Response: `Use this config instead.`
   
I can agree that my use of `initiating` and `receiving` as qualifiers all over the place doesn't help with clarity.

3. The initiating endpoint starts a series of `Connector Syncs`.
4. The initiating endpoint sends a message indicating it is done.
5. The receiving endpoint starts a series of `Connector Syncs`.
6. The receiving endpoint sends a message indicating it is done.
7. The network connection is closed.

A CLPR Endpoint's response messages for attesting to its latest configuration
are specific to that CLPR Endpoint type and are defined in that type's
extension of the `CLPR Connector Protocol`. It may be the case that a
series of state proofs showing incremental verifiable changes are necessary
to attest to the latest configuration. An explanation of how to update the
state proof verifier through these attestations of configuration must be
provided in the specification of the extension for the CLPR Endpoint type.

> 

Reviewer: This paragraph is a bit difficult to parse.  It might be clearer with some rewording, or perhaps instead of wording as "it might be needed", just declare that the proof of remote endpoint configuration supports the option to send multiple state proofs showing incremental changes if necessary.

Author: I will add a mark to rewrite when I get to this paragraph in the next major editing pass.

The protocol for a `Connector Sync` has the following pattern:

1. The initiating endpoint sends a `Proof of Connector Queue State` message
   for its local connector queue state.
    1. If the receiving endpoint does not have a matching connector, a reject
       message is sent in response and nothing further happens in the sync
       for this specific connector.
2. Otherwise, the receiving endpoint sends a `Proof of Connector Queue State`
   message for its local connector queue state.
3. The initiating endpoint sends a `Throttle Specification` which determines
   how many queued messages it is willing to receive in this sync.
4. The receiving endpoint responds with its own `Throttle Specification`.
5. Each endpoint sends a sequence of `Proof of Message Sequence` messages that
   communicate the next sequence of messages in the queue while respecting the
   throttle specifications.
6. TODO: Add Error Handling Into This Workflow
7. Each endpoint sends a `Finished With This Sync` message to indicate they are

>    ready to proceed to the next connector sync.

Reviewer: This could use a mermaid state diagram.

Author: I'll add a note to create a sequence diagram.  Although our discussion on whether or not to break this up into multiple gRPC services would impact this.

#### Abstract Channel Messages

The required content of channel messages passed between endpoints is
specified here. Each endpoint type will have its own data format and may

> provide additional data in the messages that is irrelevant. The additional

Reviewer: Do you mean to say `relevant` here?
```suggestion
provide additional data in the messages that is relevant. The additional
```

data may include metadata related to how this content is stored in the
state of each endpoint. The hashing algorithm used and the state proof
paradigm are determined by the type of endpoint sending the content.

##### Abstract Proof of Remote CLPR Endpoint Configuration

A `Proof of Remote CLPR Endpoint Configuration` Channel Message has two parts.

1. The configuration for the remote CLPR Endpoint stored in the local
   endpoint's state
2. The state proof validating the hash of the bytes of the above.

The content of the remote CLPR Endpoint configuration includes at minimum
the following:

1. `clpr_endpoint_type` : The type of the CLPR Endpoint.
2. `clpr_endpoint_id` : The CLPR Endpoint's Id

> 3. `clpr_endpoint_ips` : The list of remote endpoint ip addresses

Reviewer: These will likely be fields in a message that is CLPR specific, so we probably don't need to include `clpr` in each field name.
Also, we may wish to support addresses other than IP (e.g. FQDN or discoverable node ID).
```suggestion
1. `endpoint_type` : The type of the CLPR Endpoint.
2. `endpoint_id` : The CLPR Endpoint's Id
3. `endpoint_addresses` : The list of remote endpoint addresses
```

4. `proof_configuration` : The latest known configuration for validating
   proofs from the remote CLPR Endpoint

Note that while this is the configuration from the remote endpoint, this
content is in the local CLPR Endpoint's data format. Both the
`clpr_endpoint_id` and the `proof_configuration` are variable length byte
arrays to hold the appropriate content. There may be sub structure to the
`proof_configuration` byte array, but that substructure only matters to the
code implementing the proof verifier for the remote CLPR Endpoint's state
proofs.

##### Abstract Proof of Connector Queue State

A `Proof of Connector Queue State` Channel Message has two parts.

1. The content of the local connector's queue state
2. The state proof validating the hash of the bytes of the above content

The content of the endpoint's connector queue state is in the data format
determined by the endpoint type and consists of at least the following:

> 1. `connector_key` : The connector's public key identifier

Reviewer: The key probably needs at least a type (e.g. ECDSA, HinTS, Falcon, ED25519, etc...).
It might be reasonable to support more than just a single public key as well (e.g. something more like an Hiero `Key` message).

2. `connector_alias` : The endpoint's optional short name for the connector
3. `in_received` : The latest received incoming message's sequence number
4. `in_running_hash` : The running hash of received messages after processing

>    the `in_rec` incoming message.

Reviewer: Should this match the line two above?
```suggestion
   the `in_received` incoming message.
```

---
hip: <HIP number (assigned by the HIP editor), usually the PR number>
title: Cross-Ledger Protocol (CLPR)
author: Edward Wertz <@edward-swirldslabs>
working-group: Leemon Baird <leemon@hashgraph.com>, Richard Bair <@rbair32>, Jasper Potts <@jasperpotts>, Joseph Sinclair <@jsync-swirlds>, Rohit Sinha <@rsinha>, Kiran Pachhai <@kpachhai>, Torfin Olsen <torfinn@hashgraph.com>, Edward Wertz <@edward-swirldslabs>
requested-by: Leemon Baird <leemon@hashgraph.com>
type: "Standards Track"
category: "Core"
needs-council-approval: "Yes"
status: "Idea"
created: 2026-06-18
discussions-to:
  <A URL pointing to the official discussion thread. Ex: https://github.com/hashgraph/hedera-improvement-proposal/discussions/000>
updated: <Latest date HIP was updated, in YYYY-MM-DD format.>
requires: 1200
replaces: N/A
superseded-by: N/A
---

## Abstract

> This HIP introduces the `Cross-Ledger PRotocol` or `CLPR`, pronounced `Clipper`

Reviewer 2: Like the name much better do we need the `CLPR` at all any more so can we just say "called Clipper" rather than "pronounced Clipper"

Author: This probably needs to be discussed more between all stakeholders.

like the type of ship. `CLPR` or `Clipper` is an interledger communication
protocol that uses a connector abstraction to manage the logistics of payments

> and sending application messages between different ledgers. The use of the

Reviewer 3: ```suggestion
and sending messages between different ledgers. The use of the
```

connector abstraction allows for a clean separation of concerns between
application logic and the mechanics of sending messages via state proofs of
connector state. Combining the use of state proofs to establish message
order with the semantics of guaranteed, exactly once, in-order message
handling, Clipper is able to preserve the asynchronous Byzantine
Fault Tolerant (aBFT) security profile of an already aBFT secure ledger.

![Separable Connector and Application](../assets/hip-0000-clpr/clpr-simple-high-level-diagram.drawio.svg)

Two Clipper enabled ledgers can communicate directy with each other without
the use of intermediate relays, bridges, or oracles. If both ledgers are
aBFT, then the combined interaction is aBFT. If one ledger was
only BFT while the other was aBFT, the combined interaction is BFT. Clipper
preserves the weaker security profile of the interacting ledgers.

> Clipper is extensible in two ways. The `Clipper Application Protocol`

Reviewer 3: I'm not sure if this paragraph is useful. I am still trying to wrap my head around it after 3-4 readings.

Author: I'm modifying the paragraph.

> specifies the space of possible application use cases and can can be

Reviewer 3: ```suggestion
specifies the space of possible application use cases and can be
```

extended to with new sets of messages to enable new use cases. The
`Clipper Connector Protocol` can be extended to support new ledger types and
state proof formats.

In this HIP we specify the core behavior of a generic `Clipper Endpoint`,

> the core mechanics of a `Clipper Application`, and Hiero's and

Reviewer 3: ```suggestion
the core mechanics of a `Clipper Application`, and Hiero's
```

implementation of both.

## Glossary of Terms

* `Clipper`
    * A framework and collection of protocols for enabling interledger
      applications to communicate through aBFT secure message passing via the
      connector abstraction.
* `Clipper Application`
    * A distributed application installed on two or more ledgers that passes
      messages through a connector to communicate. A clipper application's

>       functionality is organized into one or more use cases.

Reviewer 3: The term "use case" isn't defined. But actually, I'm not sure yet that it is a useful entity to name. Maybe just leave off this sentence.

Reviewer 2: I thought the sentance " A clipper application's functionality is organized into one or more use cases." was not needed or helpful

Reviewer: Perhaps a reword?
```suggestion
      messages through a connector to communicate. A clipper application's
      functionality is supported through multiple layers of the Clipper system.
```

Author: I am modifying the text.

* `Application Message`
    * A messages sent from one application installed on one ledger to a
      remote application installed on a different ledger through a connector
      between the two ledgers.
* `Clipper Application Protocol`
    * The Clipper Application Protocol is the specification of application
      messages, callback APIs, and semantics for asynchronously handling the
      messages, all organized into discrete use cases.
* `Connector`
    * A connector manages the outgoing application message queue, permission to
      submit messages to the queue, and payment mechanics for handling sent and

>       received application messages.

Reviewer 3: Are there other types of messages than "application messages"? I like thinking of them as just "messages". They're just bytes. I don't care what's in them.

Author: The conflict in my mind was the messages that we pass on the network connection between two CLPR Endpoints.   Those messages are state proofs.  But they are also literally `message` in protobuf.

> * `Connector State Proof`

Reviewer 2: We are going to need to think about state proofs more with all my painful findings. Just how do we spec them and is there any hope of different ledgers using each other's state proofs. Also what does the state look like we are proving and how does this work with MegaMap etc. Where a leaf that you are proving is more complex than just its contents and is it a single leaf or many leaves etc.

Author: Will the new state proof details be called out explicitly in the mega map hip?

    * A state proof that establishes the state of a connector, indicating the
      messages received and sent, or provides proof of a portion of the outgoing
      message queue contents of a connector.
* `Clipper Endpoint`
    * A clipper endpoint is a network server endpoint (ip address and port)
      for a machine that can share state proofs for one or more connectors.
      There may be multiple clipper endpoints for a ledger, but there is a

>       single ledger associated with each clipper endpoint.

Reviewer 3: There is one or more connector per endpoint, but multiple endpoints per ledger -- I assume this is because each node in the ledger can run its own endpoint, but any given node would (usually?) have a single endpoint?

Reviewer 3: How do we deal with the fact that ledger endpoints could change daily, or even more frequently?

Author: Yes to the first question.  

A connection attempt to a specific endpoint id is going to round-robin through the available addresses for endpoints until it finds one that can speak the proof type it is configured to verify. 

If all of the ip addresses have changed and there is no intersection, then the configuration of the remote endpoint will need to be updated through a connector configuration update.

We probably need to have some guards.  
1. First connector registered for an CLPR Endpoint can provide configuration. 
2. After this, CLPR Endpoint configuration is only undateable through talking directly to one of the remote endpoints. 
3. Successive connector registration based updates to CLPR Endpoint are locked
4. If it ever becomes the case that a the configuration is not updatable through talking directly to the endpoints, then updates through connector registration become unlocked.

* `Clipper Endpoint Type`
    * Different types of ledgers and clipper endpoints may employ
      different state proof mechanisms or have different data formats for
      messages and other content stored in their state.
* `Clipper Endpoint Configuration`
    * The configuration for a set of clipper endpoints indicates their type, the
      ip addresses and ports to connect to, and the configuration to use for
      validating the state proofs from the endpoints. It is assumed that all
      endpoints in a configuration are for the same ledger.
* `Communication Channel`
    * Two Clipper Endpoints are able to form a communication channel between
      them when there has been at least one matching connector registered
      with the endpoints where the registration provides the configuration for

>       the remote endpoint.

Reviewer 3: Maybe it is covered later in the document, but it is critical that ledgers don't need to be able to ping each other to form a channel. One ledger might be behind a firewall. It reaches out to the other ledger and the bidirectional communication channel is formed. But it cannot be that both ledgers have to be able to dial each other up.

Author: I say it here that at least one has to specify the configuration of the other.  Meaning the other can be left off.   But I do say it explicitly elsewhere that if both are left off, then neither will be able to speak to the other.

* `Clipper Connector Protocol`
    * The Clipper Connector Protocol is the specification of state proofs for

>       conveying connector state and updating clipper endpoint configuration.

Reviewer: Here, you could use a definition list to be more screen-reader friendly.
e.g.
```html
<dl>
  <dt>Clipper</dt>
  <dd>A framework and collection of protocols for enabling interledger
  applications to communicate through aBFT secure message passing via the
  connector abstraction.</dd>
</dl>
```

>       conveying connector state and updating clipper endpoint configuration.

Reviewer 3: Does this proof also include the messages? Or does the protocol not include messages?

Author: Part of the connector state is the outgoing queue of messages.
